### PR TITLE
Debuff for hazardous materials

### PR DESCRIPTION
**Pull request by Markh3tr**
## What
Certain materials such as arsenic and asbestos shouldn't be hold in your bare hands.
This very much discourages that.

## Implementation Details
Possibility for multiple types of poisons such as inhalation and contact poison.
Applies damage to the player for holding certain variants of a material.
I'm not quite sure yet how to balance everything because lead for example is used quite often and shouldn't kill you but add some sort of other debuff.

## Outcome
Encourages players to handle dangerous compounds with protective equipment (TBA) as well as automation and using pipes to transport them.

## Additional Information
I'm planning to add config options to disable it because some players may dislike this feature.

### CODE REVIEW

**Comment by Markh3tr** [comment]
Perhaps it could also be added (with config) that entering the nether in general requires some protective equipment.

**Comment by mikerooni** [comment]
> Perhaps it could also be added (with config) that entering the nether in general requires some protective equipment.

This would be nice to have for any arbitrary dimension even - would give some nice additional opportunities to modpack creators ^^

**Comment by Ghostipedia** [comment]
> > Perhaps it could also be added (with config) that entering the nether in general requires some protective equipment.
> 
> This would be nice to have for any arbitrary dimension even - would give some nice additional opportunities to modpack creators ^^

Imo, if you're going to include this as a feature, it should be controlled with kube. 

1. A way to set a list of/single dimensions to be protected by custom item tags. Eg. Undergarden is hazardous unless the tag "blahblah:blah_protection_blah" or allow us to use a list and set tags like blah:heat_protected_blah
2. Some way to indicate the protection on the tooltip. Either by passing in a lang key ourselves or smthn
3. This would all be a lot more relevant if GTM had its armor in place (EG needing nano to enter the end etc)

Honestly the more I think about it the more this (Dimension Hazards/Armor Protection) sounds like it'd be more suited for a different mod / an Add-on rather than in the base mod, but that's my feedback if it was actually considered.

**Comment by Markh3tr** [comment]
> > > Perhaps it could also be added (with config) that entering the nether in general requires some protective equipment.
> > 
> > 
> > This would be nice to have for any arbitrary dimension even - would give some nice additional opportunities to modpack creators ^^
> 
> Imo, if you're going to include this as a feature, it should be controlled with kube.
> 
> 1. A way to set a list of/single dimensions to be protected by custom item tags. Eg. Undergarden is hazardous unless the tag "blahblah:blah_protection_blah" or allow us to use a list and set tags like blah:heat_protected_blah
> 2. Some way to indicate the protection on the tooltip. Either by passing in a lang key ourselves or smthn
> 3. This would all be a lot more relevant if GTM had its armor in place (EG needing nano to enter the end etc)
> 
> Honestly the more I think about it the more this (Dimension Hazards/Armor Protection) sounds like it'd be more suited for a different mod / an Add-on rather than in the base mod, but that's my feedback if it was actually considered.

Well for now I'm only considering general PPE because as of now heat resistance will be a thing in GT armor so that could be factored in separately.

Furthermore the restrictions on dimensions and such will all be able to be toggled in the configs because I would personally see them used more in harder modpacks rather than community pack for example.
I probably will wait until armor is done properly so the hazard protection system can hook into that neatly.

File: src/main/java/com/gregtechceu/gtceu/api/data/chemical/material/properties/HazardProperty.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            this.equipmentTypes = List.of(equipmentTypes);
+        }
+
+        public boolean isProtected(LivingEntity livingEntity){
+            List<ArmorItem.Type> correctArmorItems = new ArrayList<>();
+            for (ArmorItem.Type equipmentType: equipmentTypes){
+                ItemStack armor = livingEntity.getItemBySlot(equipmentType.getSlot());
+                if (!armor.isEmpty() && armor.getItem() instanceof ArmorComponentItem armorItem && armorItem.getArmorLogic().isPPE())
+                    correctArmorItems.add(armorItem.getType());
+            }
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/api/data/chemical/material/properties/HazardProperty.java]: this should also allow tagging any item as valid.

File: src/main/java/com/gregtechceu/gtceu/api/item/TagPrefixItem.java [hunk 2]
```diff
@@ -100,7 +104,11 @@ public void inventoryTick(ItemStack stack, Level level, Entity entity, int slotI
         if (entity instanceof LivingEntity livingEntity) {
             if (livingEntity.tickCount % 20 == 0) {
 
-                if (tagPrefix != TagPrefix.ingotHot || !material.hasProperty(PropertyKey.BLAST)) return;
+
+                if (tagPrefix != TagPrefix.ingotHot || !material.hasProperty(PropertyKey.BLAST)) { //ignore hazards for hot ingots
+                    GTUtil.applyHazardEffects(material,livingEntity,()->material.getProperty(HAZARD).getHazardType().isAffected(tagPrefix));
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/api/item/TagPrefixItem.java]: formatting please. spaces between parameters.

File: src/main/java/com/gregtechceu/gtceu/api/item/armor/IArmorLogic.java [hunk 3]
```diff
@@ -50,6 +50,8 @@ default boolean canBreakWithDamage(ItemStack stack) {
         return false;
     }
 
+    default boolean isPPE() {return false;}
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/api/item/armor/IArmorLogic.java]: formatting: don't make one-liners.

File: src/main/java/com/gregtechceu/gtceu/api/item/forge/GTBucketItem.java [hunk 4]
```diff
@@ -87,6 +98,14 @@ public int getBurnTime(ItemStack itemStack, @Nullable RecipeType<?> recipeType)
         return -1;
     }
 
+    @Override
+    public void inventoryTick(ItemStack stack, Level level, Entity entity, int slotId, boolean isSelected) {
+        super.inventoryTick(stack, level, entity, slotId, isSelected);
+        if(entity instanceof LivingEntity livingEntity && livingEntity.tickCount % 20 == 0)
+           GTUtil.applyHazardEffects(material,livingEntity,()->true);
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/api/item/forge/GTBucketItem.java]: formatting

File: src/main/java/com/gregtechceu/gtceu/common/data/GTItems.java [hunk 5]
```diff
@@ -121,14 +121,15 @@ public static void generateMaterialItems() {
                                 .properties(p -> p.stacksTo(tagPrefix.maxStackSize()))
                                 .model(NonNullBiConsumer.noop())
                                 .color(() -> TagPrefixItem::tintColor)
-                                .onRegister(GTItems::cauldronInteraction)
+
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/common/data/GTItems.java]: what? why'd you remove the dust washing?

File: src/main/java/com/gregtechceu/gtceu/common/data/GTItems.java [hunk 6]
```diff
@@ -1571,6 +1572,12 @@ public static ICustomDescriptionId cellName() {
         .onRegister(attach(ElectricStats.createElectricItem(25600000L, GTValues.IV)))
         .register();
 
+    public static ItemEntry<ComponentItem> MASK_FILTER= REGISTRATE.item("mask_filter", ComponentItem::create)
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/common/data/GTItems.java]: formatting: spaces between field and set field.

File: src/main/java/com/gregtechceu/gtceu/common/data/materials/ElementMaterials.java [hunk 7]
```diff
@@ -161,6 +166,7 @@ public static void register() {
         Chlorine = new Material.Builder(GTCEu.id("chlorine"))
                 .gas(new FluidBuilder().state(FluidState.GAS).customStill())
                 .element(GTElements.Cl)
+                .hazard(HazardProperty.HazardType.INHALATION_POISON,new HazardProperty.HazardEffect(500, MobEffects.BLINDNESS), new HazardProperty.HazardDamage(2,1), false)
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/common/data/materials/ElementMaterials.java]: formatting.

File: src/main/java/com/gregtechceu/gtceu/common/data/materials/FirstDegreeMaterials.java [hunk 8]
```diff
@@ -970,6 +976,7 @@ public static void register() {
                 .color(0xf5feff).secondaryColor(0xa4ebf1)
                 .flags(DISABLE_DECOMPOSITION)
                 .components(Sodium, 1, Oxygen, 1, Hydrogen, 1)
+                .hazard(HazardProperty.HazardType.CORROSIVE,new HazardProperty.HazardEffect(5000, MobEffects.WEAKNESS, MobEffects.DIG_SLOWDOWN), new HazardProperty.HazardDamage(2,1))
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/common/data/materials/FirstDegreeMaterials.java]: formatting.

File: src/main/java/com/gregtechceu/gtceu/data/tags/ItemTagLoader.java [hunk 9]
```diff
@@ -34,6 +34,7 @@ public static void init(RegistrateTagsProvider<Item> provider) {
         create(provider, lens, Color.Blue, GTItems.MATERIAL_ITEMS.get(lens, Sapphire).getId());
 
         create(provider, CustomTags.TAG_PISTONS, Items.PISTON, Items.STICKY_PISTON);
+        create(provider, "personal_protective_equipment",Items.DIAMOND_HELMET.kjs$getIdLocation(),Items.DIAMOND_CHESTPLATE.kjs$getIdLocation(),Items.DIAMOND_LEGGINGS.kjs$getIdLocation(),Items.DIAMOND_BOOTS.kjs$getIdLocation()); //TODO equipment
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/data/tags/ItemTagLoader.java]: don't use the KubeJS methods.

File: src/main/java/com/gregtechceu/gtceu/utils/GTUtil.java [hunk 10]
```diff
@@ -398,19 +403,43 @@ public static FluidStack getFluidFromContainer(Object ingredient) {
...
+            if (biome.warmEnoughToRain(blockPos.above()) || biome.coldEnoughToSnow(blockPos.above()))
                 return false;
-            }
-        }
-        if (world.getBiome(blockPos.above()).is(BiomeTags.IS_END)) {
+
+        if (world.getBiome(blockPos.above()).is(BiomeTags.IS_END))
             return false;
-        }
+
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/utils/GTUtil.java]: why'd you remove the brackets?? this makes it much more difficult to see where the if statement ends, especially if it's nested like this.

**Review by screret — CHANGES_REQUESTED** [review]

File: src/main/java/com/gregtechceu/gtceu/common/data/GTItems.java (hunk 5 above)
Comment (Markh3tr) [reply src/main/java/com/gregtechceu/gtceu/common/data/GTItems.java]: oopsie daisy

File: src/main/java/com/gregtechceu/gtceu/data/tags/ItemTagLoader.java (hunk 9 above)
Comment (Markh3tr) [reply src/main/java/com/gregtechceu/gtceu/data/tags/ItemTagLoader.java]: this wasn't meant to be in the commit anyway

File: src/main/java/com/gregtechceu/gtceu/api/data/chemical/material/properties/HazardProperty.java [hunk 11]
```diff
@@ -92,10 +93,11 @@ public enum ProtectionType{
 
         public boolean isProtected(LivingEntity livingEntity){
             List<ArmorItem.Type> correctArmorItems = new ArrayList<>();
-            for (ArmorItem.Type equipmentType: equipmentTypes){
+            for (ArmorItem.Type equipmentType: equipmentTypes) {
                 ItemStack armor = livingEntity.getItemBySlot(equipmentType.getSlot());
-                if (!armor.isEmpty() && armor.getItem() instanceof ArmorComponentItem armorItem && armorItem.getArmorLogic().isPPE())
-                    correctArmorItems.add(armorItem.getType());
+                if(!armor.isEmpty() && ((armor.getItem() instanceof ArmorComponentItem armorItem && armorItem.getArmorLogic().isPPE()) || armor.getTags().anyMatch(tag -> tag.location().equals(new ResourceLocation("forge", "ppe"))))){
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/api/data/chemical/material/properties/HazardProperty.java]: make the tag:
1. a constant in `CustomTags`
2. you can just do `tag.equals(CustomTags.PPE_ARMOR)`

Comment (screret) [reply src/main/java/com/gregtechceu/gtceu/api/data/chemical/material/properties/HazardProperty.java]: 3. it can be a tag in the `gtceu` namespace.

**Review by screret — CHANGES_REQUESTED** [review]

**Review by screret — APPROVED** [review]

**Comment by TheAlaskanPhysicist** [comment]
LGTM

**Pull request merged**